ceremonies/meetup-validation/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+/// The participant vecs are always vecs of participant ids
+/// The partitipant_vote and partcipant_attestations vecs (and their derived vecs) are indexed by the participant index
+/// ie. participant_votes[i] holds the vote of participant i
+
+type ParticipantIndex = usize;
+type Participants = Vec<ParticipantIndex>;
+type Attestations = Vec<Vec<ParticipantIndex>>;
+// (a, b) : a is the number of attestations
+//			b are the participants that have those number of attestations
+type ParticipantGroup = (usize, Participants);
+
+pub fn get_participant_judgements(
+	participants: &Participants,
+	participant_votes: &Vec<u32>,
+	participant_attestations: &Attestations,
+	attestation_threshold_fn: fn(usize) -> usize,
+) -> Result<ParticipantJudgements, MeetupValidationError> {
+	let mut participant_judgements =
+		ParticipantJudgements { legit: participants.clone(), excluded: vec![] };
+	participant_judgements.exclude_participants(get_excluded_participants_no_vote(
+		&participant_judgements.legit,
+		participant_votes,
+	));
+
+	let (n_confirmed, _num_votes) =
+		find_majority_vote(&participant_judgements.legit, participant_votes)?;
+
+	participant_judgements.exclude_participants(get_excluded_participants_wrong_vote(
+		&participant_judgements.legit,
+		participant_votes,
+		n_confirmed,
+	));
+
+	participant_judgements.exclude_participants(get_excluded_participants_num_attestations(
+		&participant_judgements.legit,
+		participant_attestations,
+		attestation_threshold_fn,
+	));
+
+	Ok(participant_judgements)
+}
+
+fn get_excluded_participants_no_vote(
+	participants: &Vec<ParticipantIndex>,
+	participant_votes: &Vec<u32>,
+) -> Vec<(ParticipantIndex, ExclusionReason)> {
+	// We want to get rid of all participants that did not vote (ie. have a vote of 0 (default storage value) because they did not receive any attestations).
+	// This needs to happen before we compute the majority vote, because otherwise it would be possible to receive a majority vote of 0
+	// in the case where more than half of the participants did not show up.
+
+	let mut excluded_participants: Vec<(ParticipantIndex, ExclusionReason)> = vec![];
+	for i in participants {
+		match participant_votes[*i] {
+			v if v > 0 => continue,
+			_ => excluded_participants.push((*i, ExclusionReason::NoVote)),
+		}
+	}
+	excluded_participants
+}
+
+fn get_excluded_participants_wrong_vote(
+	participants: &Participants,
+	participant_votes: &Vec<u32>,
+	n_confirmed: u32,
+) -> Vec<(ParticipantIndex, ExclusionReason)> {
+	let mut excluded_participants: Vec<(ParticipantIndex, ExclusionReason)> = vec![];
+	for i in participants {
+		if participant_votes[*i] != n_confirmed {
+			excluded_participants.push((*i, ExclusionReason::WrongVote))
+		}
+	}
+	excluded_participants
+}
+
+/// The algorithm works as follows:
+/// We find the participant with the fewest incoming/outgoing attestations
+/// If it if below the threhsold, we exclude the participant
+fn get_excluded_participants_num_attestations(
```

Comment:
I am not so happy with this function I think it does too much, why do we not separate the handling of the exclusion based on incoming and outgoing attestations?

I think it should be possible to do separate them without getting a performance overhead.

Reply:
the algorithm works by selecting always the participant with the fewest incoming OR outgoing attestations and removing it from the graph. so participant 1 could have 0 incoming attestations an gets removed. then participant 5 could have only 2 outgoing attestations and be removed. then participant 7 could have only 3 incoming and be removed. so we cannot separate the two unfortunatley.